### ISSUE TITLE
Figure title being removed by seaborn objects API when plotting on subfigures

### ISSUE THREAD

**Issue by thuiop** [mwaskom/seaborn#3211]
I recently came across an odd behaviour with the seaborn objects API when using subfigures. Here is a minimal example : 
```
import seaborn as sns
import seaborn.objects as so
import matplotlib.pyplot as plt

fig = plt.figure(constrained_layout=True)
subfigs = fig.subfigures(1,2)
tips = sns.load_dataset("tips")
p = (
    so.Plot(tips, "total_bill")
    .add(so.Bars(), so.Hist())
)
p.on(subfigs[0]).plot()

ax = subfigs[1].subplots()
ax.scatter([1],[1])

fig.suptitle("Test title")
plt.show()
```
which results in the title missing from the image :
![title_issue_bad](https://user-images.githubusercontent.com/1338337/210242982-57262fb0-d1d4-4aab-b400-8f59cae522f3.png)

Commenting the `p.on(subfigs[0]).plot()` results in the title reappearing.
I have done a bit of digging and found that changing  line 186 from the _core/subplots.py file from `figure = target.figure` to `figure = target` seems to solve the issue. Is there a specific reason to why it fetches the parent figure currently, since Subfigure is supposed to be a drop-in replacement for Figure ? I also expect this will not have the intended behaviour if we deal with subfigures of subfigures.

**Comment by mwaskom** [comment]
I can replicate but am a little confused about what's happening. Is there a reason you think that the line you called out is the culprit, or were you just poking around? If you move the suptitle text over to a coordinate like (.98, 1) you can see that it's actually still there is something being plotted over it. And yet, modifying the zorder property doesn't seem to help.

> Is there a specific reason to why it fetches the parent figure currently, since Subfigure is supposed to be a drop-in replacement for Figure ? 

Unfortunately this abstraction isn't perfect — the main methods that get called on the `Plotter._figure` object are `savefig` and `set_layout_engine`, which `SubFigure` doesn't have — I think there is a need to have a pointer to the parent figure, although maybe the behavior of `Plot.save` when it's being drawn on a subfigure is undefined. (Good point about the nested subplots edge case too).

**Comment by mwaskom** [comment]
Looks like this is replicable with pure matplotlib:

```python
f = plt.figure()
sfs = f.subfigures(1, 2)
sfs[0].subplots()
sfs[1].subplots()
f.suptitle("Test title")
```
![image](https://user-images.githubusercontent.com/315810/211165334-c97b95a9-aea6-40ab-9c03-4e75836ca0eb.png)

**Comment by mwaskom** [comment]
Reported upstream to matplotlib: https://github.com/matplotlib/matplotlib/issues/24910

**Comment by jklymak** [comment]
However to OP was using constrained_layout and in pure matplotlib it does:

```python
f = plt.figure()
sfs = f.subfigures(1, 2, layout='constrained')
sfs[0].subplots()
sfs[1].subplots()
f.suptitle("Test title")
```

![Suptitle](https://user-images.githubusercontent.com/1562854/211210932-1113b4d0-0f66-47a8-89b4-4c04dd8d05b8.png)

Does the `Plot` object call `plt.tight_layout`?  That will override constrained_layout.

**Comment by mwaskom** [comment]
Ah ok I see what's going on here. `Plot` does use the "tight" engine by default but the intended behavior is to defer to any engine that might be set on the object passed to `Plot.on`. But in fact there was a `None` value getting passed through to `Figure.set_layout_engine`, which had the effect of unsetting the constrained layout engine from the figure that was getting passed in.

This should be fixed now in #3216; here is the reprex:

![image](https://user-images.githubusercontent.com/315810/211224609-3b867fb1-90e5-4f30-a17a-28d7efa95725.png)

I still think the default behavior of matplotlib is potentially confusing but will follow up in the matplotlib issue on that.

**Comment by thuiop** [comment]
Thanks a lot for pinpointing the issue and fixing it. I temporarily fixed it in my code by adding `.layout(engine="constrained")` thanks to your remarks.
